src/NServiceBus.Core/OpenTelemetry/Metrics/ReceiveDiagnosticsBehavior.cs

```diff
@@ -18,15 +16,12 @@ public ReceiveDiagnosticsBehavior(string queueNameBase, string discriminator)
 
     public async Task Invoke(IIncomingPhysicalMessageContext context, Func<IIncomingPhysicalMessageContext, Task> next)
     {
-        context.MessageHeaders.TryGetValue(Headers.EnclosedMessageTypes, out var messageTypes);
-        var messageTypeHeader = !string.IsNullOrEmpty(messageTypes) ? messageTypes.Split(';').FirstOrDefault() : default;
-        var messageTypeName = !string.IsNullOrEmpty(messageTypeHeader) ? messageTypeHeader.Split(',').FirstOrDefault() : default;
+        context.Extensions.Set<MetricTags>(new());
 
         var tags = new TagList(new KeyValuePair<string, object>[]
```

Comment:
Ideally we could store the tag list directly but that is currently tricky because we would be boxing and unboxing all the time we get it out of the extensions. Given the discussion we had on the calls why don't we have a nested state that is called ReceiveDiagnosticsMetricTags that then carries forward the tag list and when setting the properties assigns it to the tag list that it manages? Then you can also get rid of the complicated if exists calls

We use a similar pattern in some other behaviors. Look for nested state inside behavior classes

Reply:
Here is the idea https://github.com/Particular/NServiceBus/pull/7084